crates/rover-std/src/fs.rs

```diff
@@ -54,23 +55,70 @@ impl Fs {
         C: AsRef<[u8]>,
     {
         let path = path.as_ref();
-        let contents = str::from_utf8(contents.as_ref()).with_context(|| {
-            format!(
-                "tried to write contents to {} that was invalid UTF-8",
-                &path
-            )
-        })?;
-        if !path.exists() {
-            File::create(path)
-                .with_context(|| format!("{} does not exist and it could not be created", &path))?;
-        }
-        if !path.exists() {
-            File::create(path)
-                .with_context(|| format!("{} does not exist and it could not be created", &path))?;
+        tracing::info!("checking existence of parent path in '{}'", path);
+        // Grab the parent path then attempt to canonicalize it, we can't just canonicalize the
+        // entire path because that would entail the file existing, which of course it doesn't yet.
+        let mut canonical_final_path = if let Some(parent_path) = path.parent() {
+            tracing::debug!("attempting to canonicalize parent path '{}'", parent_path);
+            Self::upsert_path_exists(parent_path)
+        } else {
+            // If the has no parent, the only conceivable case is that we are looking at the root,
+            // and as that cannot be a file we should bail out at this point
+            return Err(anyhow::anyhow!("cannot write file to root or prefix {}", path).into());
+        }?;
+
+        return if let Some(file_name) = path.file_name() {
```

Comment:
nit: seems like this would be more idiomatic as:
```
match path.file_name() {
  Some(filename) => {
    ...
    Ok(())
  }
  None => { 
    ...
    Err(...)
  }
}
```
Putting this in the return position (without the semicolon) and using a `match` statement seems closer to home.